checkout/access-checkout-android/src/test/java/com/worldpay/access/checkout/AccessCheckoutCVVTest.kt

```diff
@@ -0,0 +1,318 @@
+    }
+
+    @Test
+    fun `should not raise any events when cardValidator is not set and onUpdateCVV is called`() {
+        givenCvvIsInvalid()
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView, factory)
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = null
+
+        accessCheckoutCVV.onUpdateCVV(cvvValue)
+
+        verifyZeroInteractions(cardValidator)
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with invalid state twice - when cvv is partially valid and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsPartiallyValid()
+
+        createNewAccessCheckoutCVV().onEndUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, false)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with valid state twice - when cvv is complete and valid and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsCompletelyValid()
+
+        createNewAccessCheckoutCVV().onEndUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, true)
+    }
+
+    @Test
+    fun `should raise an onUpdate event with invalid state twice - when cvv is invalid and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsInvalid()
+
+        createNewAccessCheckoutCVV().onEndUpdateCVV(cvvValue)
+
+        verify(cardListener, times(2)).onUpdate(cvvView, false)
+    }
+
+    @Test
+    fun `should not raise any events when cardValidator is not set and onEndUpdateCVV is called`() {
+        givenCvvIsOffFocus()
+        givenCvvIsInvalid()
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView, factory)
+        accessCheckoutCVV.cardListener = cardListener
+        accessCheckoutCVV.cardValidator = null
+
+        accessCheckoutCVV.onEndUpdateCVV(cvvValue)
+
+        verifyZeroInteractions(cardValidator)
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should not raise any update events if there is no cardListener set and cvv is on focus`() {
+        givenCvvIsOnFocus()
+        givenCvvIsPartiallyValid()
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView, factory)
+        accessCheckoutCVV.cardListener = null
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should not raise any update events if there is no cardListener set and cvv is off focus`() {
+        givenCvvIsOffFocus()
+        givenCvvIsPartiallyValid()
+
+        val accessCheckoutCVV = AccessCheckoutCVV(cvvView, factory)
+        accessCheckoutCVV.cardListener = null
+        accessCheckoutCVV.cardValidator = cardValidator
+
+        verifyZeroInteractions(cardListener)
+    }
+
+    @Test
+    fun `should not raise any onUpdateLengthFilter events when cvvLengthFilter is null`() {
```

Comment:
[Can merge] - I've seen that in the code the `cvvLengthFilter` is always set when the `cardValidator` is set (due to the behaviour of the `AccessCheckoutCardDefaultFactory`). So I presume that this covers the case where the client provides their own implementation of the `CardFactory`, right?

Reply:
Yes this is correct, the client can easily provide a factory that returns a null `cvvLengthFilter` and so this test is the ensure that we can handle this.